### PR TITLE
Feature/rcp 0004 user authenication

### PR DESCRIPTION
**Pull request by MINTHANHTIKE25**
In this PR, We implemented login screen's ui and sign up screen ui. We used facebook sdk to integreate the sign in  with facebook and used the google credentail manager to implement the sign in with google . As the backend api are not ready to integrate, we didn't implement the mapper logic in domain layer and didn't implement the request and response in the data layer but we set up the ktor and room database and made the fake datasource to test the api call process. We haven't integrated the navigation logic using navigation component because we want to integrate the navigation after the onboarding team implementation's was on the main branch. We implemented the validation logic .

authScreen
login_kha
signup_kha

recipe_login_validation
recipe_valiation

### CODE REVIEW

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/LoginFragment.kt [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            btnLogin.setOnClickListener {
+                loginViewModel.onEvent(LoginFormEvent.Submit)
+            }
+
+        }
+
+
+    }
+
+    private fun textChangeEventListener() {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/LoginFragment.kt]: plz "setUpTextChangeEventListener()"

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -2,7 +2,10 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
 
+    <uses-permission android:name="android.permission.INTERNET"/>
+
     <application
+        android:name=".RecipeApps"
         android:allowBackup="true"
```
Comment (ZayPhyo01-Droid) [inline app/src/main/AndroidManifest.xml]: pls allowBackup false if it not need to backup , for more infos
https://medium.com/@jecky999/understanding-the-allowbackup-property-in-android-manifest-958151819ea

File: app/src/main/java/com/sci/recipeandroid/feature/auth/data/datasource/remote/AuthRemoteDataSourceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.sci.recipeandroid.feature.auth.data.datasource.remote
+
+class AuthRemoteDataSourceImpl(): AuthRemoteDataSource {
+    override suspend fun googleAuthentication(token:String): Result<String> {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/data/datasource/remote/AuthRemoteDataSourceImpl.kt]: authenticateWithGoogle
authenticateWithFacebook

function name should be verb

File: app/src/main/java/com/sci/recipeandroid/feature/auth/data/exception/ApiException.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package com.sci.recipeandroid.feature.auth.data.exception
+
+class ApiException(message:String,val code:Int):Exception(message)
+
+class GoogleException(message: String):Exception(message)
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/data/exception/ApiException.kt]: GoogleAuthenticationException

Comment (ZayPhyo01-Droid) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/data/exception/ApiException.kt]: GoogleException is unclear name

File: app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/FacebookAuthenticator.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+class FacebookAuthenticator {
+
+    private val callbackManager: CallbackManager = CallbackManager.Factory.create()
+
+    fun handleFacebookActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        callbackManager.onActivityResult(requestCode, resultCode, data)
+    }
+
+    fun facebookLogin(
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/FacebookAuthenticator.kt:20]: as Backend need to pass token get from Facebook SDK , we need to generate first token , so this method name should be generateFaceBookToken . Becuz this is still generate token step , not yet reach to actual fb logic step.

Comment (ZayPhyo01-Droid) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/FacebookAuthenticator.kt:20]: and google authenticator class too .

File: app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/GoogleAuthenticator.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import androidx.credentials.exceptions.GetCredentialException
+import androidx.credentials.exceptions.GetCredentialUnknownException
+import com.google.android.libraries.identity.googleid.GetGoogleIdOption
+import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
+import com.google.android.libraries.identity.googleid.GoogleIdTokenParsingException
+import com.sci.recipeandroid.feature.auth.data.exception.GoogleException
+
+class GoogleAuthenticator {
+    //TODO has to add server client id into gradle and handle exception
+    suspend fun onAuthenticate(context: Context): Result<String> {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/GoogleAuthenticator.kt]: plz check FacebookAuthenticator class

File: app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/GoogleAuthenticator.kt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            val result = credentialManager.getCredential(context, credentialRequest)
+            when (val credential = result.credential) {
+                is CustomCredential -> {
+                    if (credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
+                        try {
+                            // Use googleIdTokenCredential and extract the ID to validate and
+                            // authenticate to the server.
+                            val googleIdTokenCredential = GoogleIdTokenCredential
+                                .createFrom(credential.data)
+                            Log.d("Success", googleIdTokenCredential.idToken)
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/data/service/GoogleAuthenticator.kt]: Beware log sensitive infos

File: app/src/main/java/com/sci/recipeandroid/feature/auth/domain/repository/AuthRepository.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.sci.recipeandroid.feature.auth.domain.repository
+
+interface AuthRepository {
+    suspend fun googleAuthentication(token:String): Result<String>
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/domain/repository/AuthRepository.kt]: authenticateWithGoogle
authenticateWithFacebook

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+    }
+
+    @Deprecated("Deprecated in Java")
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        authViewModel.handleFacebookActivityResult(requestCode, resultCode, data)
+    }
+
+    private fun textDecoration() {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt]: setUp ...()

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        authViewModel.uiEvent.observe(viewLifecycleOwner) {
+            Toast.makeText(this.context, it, Toast.LENGTH_LONG).show()
+        }
+        binding.googleBtn.setOneTimeClickListener {
+            authViewModel.googleAuthentication(requireActivity())
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt]: plz inject context from koin to dependency  client class

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        _binding = FragmentAuthenticationOptionsBinding.inflate(inflater, container, false)
+        return binding.root
+    }
+
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        authViewModel.uiEvent.observe(viewLifecycleOwner) {
+            Toast.makeText(this.context, it, Toast.LENGTH_LONG).show()
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt]: this.context is nullable as
@Nullable
    public Context getContext() {
        return mHost == null ? null : mHost.getContext();
    }
    
    use requireContext

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/BaseAuthFragment.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package com.sci.recipeandroid.feature.auth.ui.screen.authentication
+
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+
+abstract class BaseAuthFragment:Fragment() {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/BaseAuthFragment.kt]: no usage found

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+class AuthOptionViewModel(
+    private val authRepository: AuthRepository,
+    private val googleAuthenticator: GoogleAuthenticator,
+    private val facebookAuthenticator: FacebookAuthenticator,
+) : ViewModel() {
+    private val _uiEvent = SingleLiveEvent<String>()
+    val uiEvent: LiveData<String> = _uiEvent
+
+    fun googleAuthentication(context: Context) {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: method name

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+                                    ?: "An Unknown Error Occurred"
+                            )
+                        }
+
+                    }
+                )
+        }
+    }
+
+    fun facebookLogin(fragment: Fragment) {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: loginFacebook or loginWithFaceBook

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+class AuthOptionViewModel(
+    private val authRepository: AuthRepository,
+    private val googleAuthenticator: GoogleAuthenticator,
+    private val facebookAuthenticator: FacebookAuthenticator,
+) : ViewModel() {
+    private val _uiEvent = SingleLiveEvent<String>()
+    val uiEvent: LiveData<String> = _uiEvent
+
+    fun googleAuthentication(context: Context) {
+        viewModelScope.launch(Dispatchers.IO) {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: IO handling should be in corresponding layer , eg data layer or service

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    fun handleFacebookActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        facebookAuthenticator.handleFacebookActivityResult(
+            resultCode = resultCode,
+            requestCode = requestCode,
+            data = data
+        )
+
+    }
+    private fun facebookAuthentication(accessToken: String) {
+        viewModelScope.launch(Dispatchers.IO) {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: plz move IO handling to corresponding layer

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/LoginViewModel.kt [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                            emailError = emailResult.errorMessage,
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    private fun logIn(email: String, password: String){
+        viewModelScope.launch(Dispatchers.IO) {
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/LoginViewModel.kt:54]: IO handling

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/SignUpViewModel.kt [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+        } else {
+            postToServer(
+                name = currentState.displayName,
+                email = currentState.email,
+                password = currentState.password
+            )
+        }
+    }
+
+    private fun postToServer(
```
Comment (ZayPhyo01-Droid) [inline app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/SignUpViewModel.kt]: signUp

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (MINTHANHTIKE25) [reply app/src/main/AndroidManifest.xml]: setting the allow back up to false as we don't want to integrate the back up 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/LoginFragment.kt (hunk 1 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/LoginFragment.kt]: changed the name of this function at 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/data/datasource/remote/AuthRemoteDataSourceImpl.kt (hunk 3 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/data/datasource/remote/AuthRemoteDataSourceImpl.kt]: change the name of this function 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt (hunk 10 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt]: inject this context using the custom context 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt (hunk 11 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/AuthOptionFragment.kt]: used the requirecontext 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/SignUpViewModel.kt (hunk 18 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/SignUpViewModel.kt]: changed this function name 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/LoginViewModel.kt (hunk 17 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/LoginViewModel.kt:54]: moved this io handling at the datasource 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt (hunk 16 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: move io handling to datasource 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt (hunk 15 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: moved this io handling to datasource 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt (hunk 13 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/viewmodel/AuthOptionViewModel.kt]: changed this function name at 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

File: app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/BaseAuthFragment.kt (hunk 12 above)
Comment (MINTHANHTIKE25) [reply app/src/main/java/com/sci/recipeandroid/feature/auth/ui/screen/authentication/BaseAuthFragment.kt]: delete this class 9baf80468f57d78d6c4b13f545e1ebaa10f9d59c

**Pull request merged**